File: openharmony/arkui_ace_engine/frameworks/bridge/arkts_frontend/koala_projects/arkoala-arkts/libarkts/src/es2panda.ts
/*
 * Copyright (c) 2022-2025 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as fs from "node:fs"
import * as path from "node:path"
import { global } from "./arkts-api/static/global"
import { Command } from "commander"
import { isNumber, throwError, withWarning, filterSource } from "./utils"
import { Es2pandaContextState } from "./generated/Es2pandaEnums"
import { AstNode, Config, Context, EtsScript, proceedToState } from "./arkts-api"

function parseCommandLineArgs() {
    const commander = new Command()
        .option('--file, <char>', 'Path to file to be compiled')
        .option('--arktsconfig, <char>', 'Path to arkts configuration file')
        .option('--ets-module', 'Do nothing, legacy compatibility')
        .option('--output, <char>', 'The name of result file')
        .option('--dump-plugin-ast', 'Dump ast before and after each plugin')
        .option('--restart-stages', 'Restart the compiler to proceed to next stage')
        .option('--generate-decl', 'Generate declaration files')
        .option('--decl-arktsconfig, <char>', 'Path to arkts configutation file for the binary compilation')
        .parse(process.argv)
        .opts()

    const restOptions = commander.args

    const filePath = path.resolve(commander.file)
    const configPath = path.resolve(commander.arktsconfig)
    const outputPath = path.resolve(commander.output)
    if (!fs.existsSync(filePath)) {
        throw new Error(`File path doesn't exist: ${filePath}`)
    }
    if (!fs.existsSync(configPath)) {
        throw new Error(`Arktsconfig path doesn't exist: ${configPath}`)
    }

    const dumpAst = commander.dumpPluginAst ?? false
    const restartStages = commander.restartStages ?? false
    const generateDecl = commander.generateDecl ?? false
    const dConfigPath = path.resolve(commander.declArktsconfig ?? commander.arktsconfig)

    return { filePath, configPath, outputPath, dumpAst, restartStages, generateDecl, dConfigPath }
}

function insertPlugin(plugin: (ast: AstNode) => void, state: Es2pandaContextState, dumpAst: boolean): AstNode {
    proceedToState(state)
    const script = EtsScript.fromContext()
    if (script === undefined) {
        throwError(`Failed to receive ast from es2panda`)
    }

    if (dumpAst) {
        console.log(`BEFORE ${stateName(state)}:`)
        console.log(filterSource(script.dumpSrc()))
    }

    const transform = plugin
    transform?.(script)

    if (dumpAst) {
        console.log(`AFTER ${stateName(state)}:`)
        console.log(filterSource(script.dumpSrc()))
    }

    setAllParents(script)
    return script
}

function restartCompiler() {
    console.log("restarting")
    const ast = EtsScript.fromContext()
    const srcText = filterSource(ast.dumpSrc())
    global.es2panda._DestroyContext(global.context)
    global.compilerContext = Context.createFromString(srcText)
}

function restartCompilerWithConfig(configPath: string, filePath: string, stdlib: string, outputPath: string) {
    console.log("restarting with another config")
    const srcText = filterSource(EtsScript.fromContext().dumpSrc())
    global.es2panda._DestroyContext(global.context)
    global.es2panda._DestroyConfig(global.config)
    global.config = Config.create([
        '_',
        '--arktsconfig',
        configPath,
        filePath,
        '--extension',
        'sts',
        '--stdlib',
        stdlib,
        '--output',
        outputPath
    ]).peer
    global.compilerContext = Context.createFromString(srcText)
}

const defaultPandaSdk = "../../../incremental/tools/panda/node_modules/@panda/sdk"
function invokeWithPlugins(
    configPath: string,
    filePath: string,
    outputPath: string,
    pluginsByState: Map<Es2pandaContextState, ((ast: AstNode) => void)[]>,
    dumpAst: boolean,
    restart: boolean,
    generateDecl: boolean,
    dConfigPath: string,
): void {
    const source = fs.readFileSync(filePath).toString()
    const sdk = process.env.PANDA_SDK_PATH ?? withWarning(
        defaultPandaSdk,
        `PANDA_SDK_PATH not set, assuming ${defaultPandaSdk}`
    )
    const stdlib = `${sdk}/ets/stdlib`

    global.filePath = filePath
    global.config = Config.create([
        '_',
        '--arktsconfig',
        configPath,
        filePath,
        '--extension',
        'sts',
        '--stdlib',
        stdlib,
        '--output',
        outputPath
    ]).peer
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    global.compilerContext = Context.createFromString(source)

    console.log("PLUGINS: ", pluginsByState.size, pluginsByState)

    if (pluginsByState.size == 0) {
        proceedToState(Es2pandaContextState.ES2PANDA_STATE_BIN_GENERATED)
        return
    }

    pluginsByState.get(Es2pandaContextState.ES2PANDA_STATE_PARSED)?.forEach(plugin => {
        insertPlugin(plugin, Es2pandaContextState.ES2PANDA_STATE_PARSED, dumpAst)
        if (restart) {
            restartCompiler()
        }
    })

    pluginsByState.get(Es2pandaContextState.ES2PANDA_STATE_CHECKED)?.forEach(plugin => {
        insertPlugin(plugin, Es2pandaContextState.ES2PANDA_STATE_CHECKED, dumpAst)
        // TODO: Normally we need just to proceed to a given state,
        //  but the compiler crashes now, so we restart
        if (restart) {
            restartCompiler()
        }
    })

    if (configPath !== dConfigPath) {
        proceedToState(Es2pandaContextState.ES2PANDA_STATE_PARSED)
        restartCompilerWithConfig(dConfigPath, filePath, stdlib, outputPath)
    }

    if (generateDecl) {
        proceedToState(Es2pandaContextState.ES2PANDA_STATE_PARSED)
        generateDeclFromCurrentContext(filePath)
    }

    proceedToState(Es2pandaContextState.ES2PANDA_STATE_BIN_GENERATED)
}

function generateDeclFromCurrentContext(filePath: string) {
    const ext = path.extname(filePath)
    const dir = path.basename(path.dirname(filePath))

    let declFile = ""
    if (dir == "sts") { // temporary workaround (sts dir -> dsts dir)
        fs.mkdirSync(path.join(path.dirname(path.dirname(filePath)), 'dsts'), { recursive: true })
        declFile = path.join(path.dirname(path.dirname(filePath)), 'dsts', path.basename(filePath))
    } else { // extension change (.sts -> .d.sts, .ets -> .d.ets)
        declFile = path.join(path.dirname(filePath), `${path.basename(filePath, ext)}.d${ext}`)
    }
    console.log(`Emiting declaration file to`, declFile)

    let out = filterSource(global.compilerContext.program.astNode.dumpSrc())

    // TODO: exports are lost on visiting
    const exportAstNodes = ["class", "abstract class", "interface", "@interface", "type"]
    exportAstNodes.forEach((astNodeText) => {
            out = out.replaceAll(`\n${astNodeText}`, `\nexport ${astNodeText}`)
        }
    )

    fs.writeFileSync(declFile,
        out
        .replaceAll("\nexport abstract class ETSGLOBAL", "\nabstract class ETSGLOBAL")
    )
}

function setAllParents(ast: AstNode) {
    global.es2panda._AstNodeUpdateAll(global.context, ast.peer)
}

function loadPlugin(configDir: string, jsonPlugin: any) {
    const pluginPath = jsonPlugin.transform ?? throwError(`arktsconfig plugins objects should specify transform`)
    /** TODO: read and pass plugin options */
    const plugin = (pluginPath.startsWith(".") || pluginPath.startsWith("/")) ?
        path.resolve(configDir, pluginPath) : pluginPath

    const pluginFunction: (config?: any) => any = require(plugin).default(jsonPlugin)
    return pluginFunction
}

function selectPlugins(configDir: string, plugins: any[], stage: string): ((arg: AstNode) => AstNode)[] | undefined {
    const selected = plugins
        .filter(it => (it.stage == stage))
        .map(it => loadPlugin(configDir, it))
    if (selected.length == 0) {
        return undefined
    }
    return selected
}

function stateName(value: Es2pandaContextState): string {
    return Es2pandaContextState[value].substring("ES2PANDA_STATE_".length)
}

function readAndSortPlugins(configDir: string, plugins: any[]) {
    const pluginsByState = new Map<Es2pandaContextState, ((ast: AstNode) => void)[]>()

    Object.values(Es2pandaContextState)
        .filter(isNumber)
        .forEach(it => {
            const selected = selectPlugins(configDir, plugins, stateName(it).toLowerCase())
            if (selected) pluginsByState.set(
                it,
                selected
            )
        })

    return pluginsByState
}

export function main() {
    const { filePath, configPath, outputPath, dumpAst, restartStages, generateDecl, dConfigPath } = parseCommandLineArgs()
    const arktsconfig = JSON.parse(fs.readFileSync(configPath).toString())
    const configDir = path.dirname(configPath)
    const compilerOptions = arktsconfig.compilerOptions ?? throwError(`arktsconfig should specify compilerOptions`)
    const plugins = compilerOptions.plugins ?? []

    const pluginsByState = readAndSortPlugins(configDir, plugins)

    invokeWithPlugins(configPath, filePath, outputPath, pluginsByState, dumpAst, restartStages, generateDecl, dConfigPath)
}

main()

File: openharmony/arkcompiler_ets_frontend/ets2panda/es2panda.cpp
/**
 * Copyright (c) 2021-2025 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "compiler/core/compilerImpl.h"
#include "generated/diagnostic.h"
#include "public/public.h"
#include "util/diagnostic.h"
#include "util/generateBin.h"

#include "es2panda.h"

namespace ark::es2panda {
constexpr size_t DEFAULT_THREAD_COUNT = 2;

namespace util {
class Options;
}  // namespace util

template <class T>
T DirName(T const &path, T const &delims = ark::os::file::File::GetPathDelim())
{
    std::size_t pos = path.find_last_of(delims);
    if (pos == std::string::npos) {
        return "./";
    }

    if (pos == 0) {
        return delims;
    }

    std::string_view dirPath = path.substr(0, pos);
    if (dirPath.compare(".") == 0 || dirPath.compare("..") == 0) {
        return path.substr(0, pos + 1);
    }

    return path.substr(0, pos);
}

SourceFile::SourceFile(std::string_view fn, std::string_view s) : filePath(fn), fileFolder(DirName(fn)), source(s) {}

SourceFile::SourceFile(std::string_view fn, std::string_view s, bool m)
    : filePath(fn), fileFolder(DirName(fn)), source(s), isModule(m)
{
}
SourceFile::SourceFile(std::string_view fn, std::string_view s, bool m, std::string_view d)
    : filePath(fn), fileFolder(DirName(fn)), source(s), isModule(m), dest(d)
{
}

SourceFile::SourceFile(std::string_view fn, std::string_view s, std::string_view rp, bool m, bool d)
    : filePath(fn),
      fileFolder(DirName(fn)),
      source(s),
      resolvedPath(DirName(rp)),
      isModule(m),
      isDeclForDynamicStaticInterop(d)
{
}

Compiler::Compiler(ScriptExtension ext) : Compiler(ext, DEFAULT_THREAD_COUNT, {}) {}

Compiler::Compiler(ScriptExtension ext, size_t threadCount) : Compiler(ext, threadCount, {}) {}

Compiler::Compiler(ScriptExtension ext, size_t threadCount, std::vector<util::Plugin> &&plugins)
    : plugins_(std::move(plugins)), compiler_(new compiler::CompilerImpl(threadCount, &plugins_)), ext_(ext)
{
}

Compiler::~Compiler()
{
    delete compiler_;
}

pandasm::Program *Compiler::Compile(const SourceFile &input, const util::Options &options,
                                    util::DiagnosticEngine &diagnosticEngine, uint32_t parseStatus)
{
    public_lib::Context context;
    ArenaAllocator allocator(SpaceType::SPACE_TYPE_COMPILER, nullptr, true);
    context.allocator = &allocator;
    context.compilingState = public_lib::CompilingState::SINGLE_COMPILING;

    try {
        return compiler_->Compile(compiler::CompilationUnit {input, options, parseStatus, ext_, diagnosticEngine},
                                  &context);
    } catch (const util::ThrowableDiagnostic &e) {
        error_ = e;
        return nullptr;
    }
}

unsigned int Compiler::CompileM(std::vector<SourceFile> &inputs, util::Options &options,
                                util::DiagnosticEngine &diagnosticEngine, std::vector<pandasm::Program *> &result)
{
    public_lib::Context context;
    context.transitionMemory =
        new public_lib::TransitionMemory(new ArenaAllocator(SpaceType::SPACE_TYPE_COMPILER, nullptr, true));
    context.allocator = context.transitionMemory->PermanentAllocator();

    context.compilingState = public_lib::CompilingState::MULTI_COMPILING_INIT;
    unsigned int overallRes = 0;
    for (auto &input : inputs) {
        try {
            options.SetOutput(std::string(input.dest));
            LOG_IF(options.IsListFiles(), INFO, ES2PANDA)
                << "> es2panda: compiling from '" << input.filePath << "' to '" << input.dest << "'";
            auto program =
                compiler_->Compile(compiler::CompilationUnit {input, options, 0, ext_, diagnosticEngine}, &context);
            result.push_back(program);
        } catch (const util::ThrowableDiagnostic &err) {
            overallRes |= 1U;
            diagnosticEngine.Log(err);
        }
        context.compilingState = public_lib::CompilingState::MULTI_COMPILING_FOLLOW;
    }
    delete context.transitionMemory;
    return overallRes;
}

std::string Compiler::GetPhasesList() const
{
    return compiler::CompilerImpl::GetPhasesList(ext_);
}

void Compiler::DumpAsm(const pandasm::Program *prog)
{
    compiler::CompilerImpl::DumpAsm(prog);
}

util::DiagnosticEngine *g_diagnosticEngine = nullptr;

}  // namespace ark::es2panda
